### ISSUE TITLE
Rename package from main to chq

### ISSUE THREAD

**Issue by jimbojetlag** [aktungmak/chq#2]
Is this an application or a library? I'm trying to extend this and add my own filter, but the package is called main.

**Comment by jimbojetlag** [comment]
Actually, I cannot add a new filter outside of the package, as variables like `TsNode.input` are not exported.

**Comment by jimbojetlag** [comment]
Sorry to bump on this issue, would it be possible to have a minimal example of how to make a pipeline from a file input to a simple dummy filter that prints something like the number of packets? I could'd find a way of making this work programmatically (after changing the source code to export required variables).

**Comment by aktungmak** [comment]
Its designed to be a compete application, so thats why I kept all the code in package main. It might make sense to split it out into individual packages for the nodes, one for the http server and router, and then bring them together in a simpler package main... but I haven't got this far yet.

As for the TsNode, you don't need to access the `.input` field since it is accessed via the `GetInputChan()` method of the `Routeable{}` interface - check out the file Router.go.

If you wanted to just route a FileInput and see a count of number of packets per PID, modify the file `basic.chq` (which is the config file, hard coded in main.go at the moment) to something like this:

```
node inp ::= FileInput mpts.ts
node ctr ::= PidCounter

conn inp -> ctr
```

Once the application is running, you can open your web browser and navigate to http://localhost:10101/ and you can see the summary of all the nodes and their status. Check out the file Server.go to see all the other endpoints, which allow you to create new nodes and connect/disconnect them. 

I am still working out the structure of some aspects of things at the moment, so this isn't as well packaged as it should be!

**Comment by jimbojetlag** [comment]
Thanks, my question was how to make a pipeline **programmatically**.

I want to read a ts file, filter some packets based on some criteria, join the filtered packet payloads to form a buffer, and then write that buffer to a file. So far I could come up with this:

``` go
node1 := NewNode1()
node1.Active = true // if not active, we get a deadlock error!

node2 := newNode2()
node2.Active = true
node1.RegisterListener(node2.GetInputChan()) // connect node1 to node2

out := make(chan TsPacket)
node2.RegisterListener(out)

// drain the out channel.
for _ = range out {
}
```

**Comment by aktungmak** [comment]
Ah I see! Sorry, I misunderstood. Sure, your method will work there. However, I was anticipating most interactions would take place using the REST API defined in Server.go, which instantiate the nodes, connect them and then bring them Active (using the `.Toggle()` method). 

For the example you specified, the FileInput and FileOutput node types will do what you need, and then the various filters already available can be used, or another type which conforms to the `Routeable{}` interface which you define.

**Comment by aktungmak** [comment]
PS the nodes are "leaky" by design to prevent pipelines from backing up. This means that if a node has no listeners it will still carry on processing. To make sure no packets get dropped, only `Toggle()` the nodes in a pipeline once they are all connected.

**Comment by jimbojetlag** [comment]
Thanks! A couple of questions about the design:
1. If i understand this correctly, each node lives in its own goroutine. When the TS source is a file on a disk, our data streaming pipeline is bound to disk IO. So even if we have multiple processes running in parallel, they have to wait for disk IO. In this case, what's the benefit of a node per goroutine? Also When the nodes are connected in series, doesn't one have to wait for the previous one to finish before it starts the processing? How does concurrency help with this? (I might have gone the design wrong, that's why I'm asking this question!).
2. I like the current pipeline design, but it does not follow the standard `io.Reader` interface. Can you see a way of doing this?

**Comment by aktungmak** [comment]
Yep thats right, each node is its own goroutine. In the case of a single chain of nodes, there is no benefit from this, since they all rely on the node before. 
The benefit comes when you want to do multiple things with the stream at the same time. For example, you can connect multiple PidFilters to the output of your FileInput or NetInput and do the processing of each PID independently of the others. I was also envisioning a use case where we have several (possibly unrelated) pipelines all being managed by one instance of chq.

With regards to 2. I had not thought about that, but I am not sure what it would look like. Do you mean that the Routable interface (which all nodes must implement) should also embed the io.Reader interface?

**Comment by jimbojetlag** [comment]
I see.

About 2, I meant connecting the filters via `io.Reader`, which is more idiomatic in go, although I'm not sure how this would fit with channels.

Have a look at https://github.com/32bitkid/mpeg design.

**Comment by jimbojetlag** [comment]
In a scenario where PIDs are not known in advance, can this design support 'dynamically' creating output nodes, inside a node that discovers the PIDs? Would it be safe to create those goroutines inside one goroutine and then toggle `Active`?

**Comment by aktungmak** [comment]
Yes, the REST API will provide details of all the nodes. So, a PatParser, PmtParser and PidCounter could all be used by client software to “discover” the contents of a stream. The client software can then instantiate and connect nodes via the REST API to build up a reactive processing chain.
